Add amdgpu target

Flakebi opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->

Add amdgpu target to rustc and enable the LLVM target.

Fix compiling `core` with the amdgpu:
The amdgpu backend makes heavy use of different address spaces. This
leads to situations, where a pointer in one addrspace needs to be casted
to a pointer in a different addrspace. `bitcast` is invalid for this
case, `addrspacecast` needs to be used.

Fix compilation failures that created bitcasts for such cases by
creating pointer casts (which creates an `addrspacecast` under the hood)
instead.

MCP: https://github.com/rust-lang/compiler-team/issues/823
Tracking issue: #135024
Kinda related to the original amdgpu tracking issue #51575 (though that one has been closed for a while).

jieyouxu commented:
r? jieyouxu

workingjubilee commented:
cc @eddyb Hello, tagging you for domain expertise if you want to chime in.

jieyouxu commented:
@rustbot author

bjorn3 commented on compiler/rustc_target/src/spec/targets/amdgcn_amd_amdhsa.rs:
```diff
@@ -0,0 +1,41 @@
+            no_builtins: true,
+
+            // Needed to use `cdylib` crate type.
+            dynamic_linking: true,
```
```suggestion
            dynamic_linking: true,
            only_cdylib: true,
            executables: false,
```

bjorn3 commented on compiler/rustc_target/src/spec/targets/amdgcn_amd_amdhsa.rs:
```diff
@@ -0,0 +1,41 @@
+
+            // amdgpu backend does not support libcalls.
+            no_builtins: true,
+
```
```suggestion
            simd_types_indirect: false,
```

assuming that no target features that can be enabled for individual functions or crates change the vector ABI.

Flakebi commented:
Thank you for the quick review!

I opened an MCP here: https://github.com/rust-lang/compiler-team/issues/823

bors commented:
:umbrella: The latest upstream changes (presumably #134822) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Flakebi committed:
Add amdgpu target

Add target and compile the amdgpu llvm backend.

Flakebi commented:
I removed changes apart from adding the target and enabling the LLVM backend from this PR. They will go into their own PRs.
I extended the platform-support documentation a bit.

There is now a tracking issue: #135024

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12583831828/job/35072237464) [(plain)](https://github.com/rust-lang/rust/commit/c47a2c9f18c9dd47ee6141368f59fdd498871805/checks/35072237464/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 26.0s done
#21 DONE 39.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.90s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
##[group]Testing stage2 book rustc (x86_64-unknown-linux-gnu)
STDOUT:

running 1 test
test /checkout/src/doc/rustc/src/platform-support/amdgcn-amd-amdhsa.md - Tier__3::Building_Rust_programs (line 34) ... FAILED
failures:


---- /checkout/src/doc/rustc/src/platform-support/amdgcn-amd-amdhsa.md - Tier__3::Building_Rust_programs (line 34) stdout ----
error[E0557]: feature has been removed
##[error] --> /checkout/src/doc/rustc/src/platform-support/amdgcn-amd-amdhsa.md:36:12
3 | #![feature(abi_amdgpu_kernel)]
  |            ^^^^^^^^^^^^^^^^^ feature has been removed

error: unwinding panics are not supported without std
error: unwinding panics are not supported without std
  |
  = help: using nightly cargo, use -Zbuild-std with panic="abort" to avoid unwinding
  = note: since the core library is usually precompiled with panic="unwind", rebuilding your crate with panic="abort" may not be enough to fix the problem

error[E0703]: invalid ABI: found `amdgpu-kernel`
##[error]  --> /checkout/src/doc/rustc/src/platform-support/amdgcn-amd-amdhsa.md:45:12
   |
12 | pub extern "amdgpu-kernel" fn kernel(/* Arguments */) {
   |            ^^^^^^^^^^^^^^^ invalid ABI
   |
   = note: invoke `rustc --print=calling-conventions` for a full list of supported calling conventions
error[E0601]: `main` function not found in crate `rust_out`
error[E0601]: `main` function not found in crate `rust_out`
##[error]  --> /checkout/src/doc/rustc/src/platform-support/amdgcn-amd-amdhsa.md:47:2
14 | }
14 | }
   |  ^ consider adding a `main` function to `/checkout/src/doc/rustc/src/platform-support/amdgcn-amd-amdhsa.md`
error: aborting due to 4 previous errors

Some errors have detailed explanations: E0557, E0601, E0703.
For more information about an error, try `rustc --explain E0557`.
For more information about an error, try `rustc --explain E0557`.
Couldn't compile the test.

failures:
    /checkout/src/doc/rustc/src/platform-support/amdgcn-amd-amdhsa.md - Tier__3::Building_Rust_programs (line 34)
test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.04s




Command CFG_RELEASE_CHANNEL="nightly" RUSTC_BOOTSTRAP="1" RUSTC_STAGE="2" RUSTC_SYSROOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" RUSTDOC_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTDOC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" RUST_TEST_THREADS="16" "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/src/doc/rustc/src/platform-support/amdgcn-amd-amdhsa.md" "--test-args" "" (failure_mode=DelayFail) has failed. Rerun with -v to see more details.
  local time: Thu Jan  2 15:00:53 UTC 2025
  network time: Thu, 02 Jan 2025 15:00:53 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #135074) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Flakebi commented:
Rebased to fix conflicts and fix tests, no changes to the target.

Flakebi commented:
Extend target documentation.
I found out that HIP can be used to load/run amdgpu binaries as well (in addition to HSA), and it works on Linux and Windows!
[Diff](https://github.com/rust-lang/rust/compare/a65bf7a56d15c49eff083527af44fd80bea86d99..f0ebf2c3c0aa79d7f112a53d4d46f7bd93818a71)

workingjubilee commented:
oh, that's awesome! that should make testing this much easier.

workingjubilee commented:
https://github.com/rust-lang/rust/pull/135030 first, and then this should land when that flag has been included for the target.

workingjubilee commented:
Right, it's in. Please rebase and include a test that specifically verifies the target doesn't work unless `-Ctarget-cpu` has been set.

Flakebi commented:
Thanks! I rebased and added a second commit for requiring `-Ctarget-cpu` (to make it easier to review the new changes, if you prefer, I can squash them). Thanks for the reminder about adding tests, I probably would have forgotten otherwise. Definitely makes sense to have them.

workingjubilee commented:
Thanks!

...uh, I have a question now, hm.

`-Ctarget-cpu` is going to show up in the metadata for our rlibs, right? So that we won't link ones with mismatching CPUs together? I felt confident about this a second ago but then I inspected the thought and I'm not entirely sure how the invalidation logic actually goes, since it would make sense for it to be attemptable usually.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13064130877/job/36453388154) [(plain)](https://github.com/rust-lang/rust/commit/1f0a21b2b62a671a6a9bd25d0e71216c8d750c11/checks/36453388154/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#22 exporting to docker image format
#22 sending tarball 28.9s done
#22 DONE 35.3s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
##[endgroup]
##[group]Testing stage2 compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu)

running 18391 tests
   [ui] tests/ui/amdgpu-require-explicit-cpu.rs#cpu ... FAILED
20%  --  3679/18391,  3623 passed, 1 failed, 55 ignored
30%  --  5518/18391,  5452 passed, 1 failed, 65 ignored
40%  --  7357/18391,  7282 passed, 1 failed, 74 ignored
50%  --  9196/18391,  9109 passed, 1 failed, 86 ignored
---


failures:

---- [ui] tests/ui/amdgpu-require-explicit-cpu.rs#cpu stdout ----

error in revision `cpu`: test compilation failed although it shouldn't!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/amdgpu-require-explicit-cpu.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "cpu" "--check-cfg" "cfg(test,FALSE,nocpu,cpu)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/amdgpu-require-explicit-cpu.cpu" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type=cdylib" "--target=amdgcn-amd-amdhsa" "-Ctarget-cpu=gfx900"
--- stderr -------------------------------
error: linker `rust-lld` not found
   |
   = note: No such file or directory (os error 2)
   = note: No such file or directory (os error 2)

error: aborting due to 1 previous error
------------------------------------------



failures:
    [ui] tests/ui/amdgpu-require-explicit-cpu.rs#cpu
test result: FAILED. 18199 passed; 1 failed; 191 ignored; 0 measured; 0 filtered out; finished in 179.16s

Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
Build completed unsuccessfully in 0:18:27
```

</details>

Flakebi commented:
> `-Ctarget-cpu` is going to show up in the metadata for our rlibs, right? So that we won't link ones with mismatching CPUs together? I felt confident about this a second ago but then I inspected the thought and I'm not entirely sure how the invalidation logic actually goes, since it would make sense for it to be attemptable usually.

It seems like the target-cpu does not show up in the metadata (or at least I didn’t find it when opening the rmeta file with a text editor).
Linking a library compiled for one target-cpu to a library from another target-cpu works without complaining (it even works when using `-Clinker-plugin-lto` where lld merges the bitcode files, LLVM doesn’t complain about this).
I guess in many cases this is undesirable (calling a gfx900 function from a gfx1030 function won’t work well), but for some combinations it should be ok (like compiling a library for gfx10-1-generic and using it from a library compiled for gfx1011).

tgross35 commented:
Maybe this is a case that [target modifiers](https://github.com/rust-lang/rfcs/pull/3716) will help prevent?

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13064598731/job/36454667531) [(plain)](https://github.com/rust-lang/rust/commit/2e407f268ac0b6ed4e6d6837ed656d353bbc962a/checks/36454667531/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 28.2s done
#21 DONE 49.7s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
10%  --  54/530,  40 passed, 0 failed, 14 ignored
20%  -- 107/530,  83 passed, 0 failed, 24 ignored
30%  -- 160/530, 128 passed, 0 failed, 32 ignored
40%  -- 213/530, 181 passed, 0 failed, 32 ignored
   [assembly] tests/assembly/targets/targets-amdgpu.rs#amdgcn_amd_amdhsa ... FAILED
60%  -- 319/530, 286 passed, 1 failed, 32 ignored
70%  -- 372/530, 339 passed, 1 failed, 32 ignored
80%  -- 424/530, 391 passed, 1 failed, 32 ignored
90%  -- 477/530, 444 passed, 1 failed, 32 ignored
90%  -- 477/530, 444 passed, 1 failed, 32 ignored
100% -- 530/530, 487 passed, 1 failed, 42 ignored


failures:

---- [assembly] tests/assembly/targets/targets-amdgpu.rs#amdgcn_amd_amdhsa stdout ----

error in revision `amdgcn_amd_amdhsa`: compilation failed!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/assembly/targets/targets-amdgpu.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--cfg" "amdgcn_amd_amdhsa" "--check-cfg" "cfg(test,FALSE,amdgcn_amd_amdhsa)" "-O" "-Cdebug-assertions=no" "--emit" "asm" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/targets/targets-amdgpu.amdgcn_amd_amdhsa/targets-amdgpu.s" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target" "amdgcn-amd-amdhsa"
--- stderr -------------------------------
--- stderr -------------------------------
error: target requires explicitly specifying a cpu with `-C target-cpu`
error: aborting due to 1 previous error
------------------------------------------




failures:
    [assembly] tests/assembly/targets/targets-amdgpu.rs#amdgcn_amd_amdhsa
test result: FAILED. 487 passed; 1 failed; 42 ignored; 0 measured; 0 filtered out; finished in 2.27s

Some tests failed in compiletest suite=assembly mode=assembly host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
Build completed unsuccessfully in 0:18:47
```

</details>

workingjubilee commented:
Indeed, the `-Ctarget-cpu` is effectively such a modifier for this target. I would prefer that we fail-closed rather than fail-open for linkage we are unsure about.

workingjubilee commented:
That said, if it's necessary to land the target first before we can realistically integrate more code that does appropriate checks, I'm fine with that.

Flakebi commented:
I would prefer landing the target first. There are more pieces depending on that and requiring consistent target-cpus sounds like something that requires a more thought out design.

workingjubilee commented:
alrighty then!

@bors r+ rollup

bors commented:
:pushpin: Commit 56795fb77a69bc61836537175564455067adada6 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 56795fb77a69bc61836537175564455067adada6 -->
<!-- homu: {"type":"Approved","sha":"56795fb77a69bc61836537175564455067adada6","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors rollup=never

workingjubilee commented:
This is blocking other related PRs from proceeding, so

@bors p=1

workingjubilee commented:
Reason for the `rollup=never`:

_@klensy in https://github.com/rust-lang/rust/issues/136399#issuecomment-2629346904_:
> #134740 adds llvm target, probably increasing libllvm size, should be rollup=never?

_@matthiaskrgr in https://github.com/rust-lang/rust/issues/136399#issuecomment-2629354902_:
> maybe we need exponentially more memory during lto or something with the new target...
>
> if've iffyd https://github.com/rust-lang/rust/pull/136309 and never'd https://github.com/rust-lang/rust/pull/134740 let's see if that help..
>
> notable though that none of the other rollup=never prs have landed in the meantime, maybe gh is just having a bad weekend :/

I think that's unlikely, but LLVM does love to randomly have a runaway compile time event, so it's probably just as well.

workingjubilee commented:
Also, if this is suspected of being the reason for failures of rollups, it should just be put through a try-job anyways, since we have enough high-risk jobs in the queue already, we don't need to send something through with only middling confidence it will work:

@bors r-
@bors try

bors commented:
:hourglass: Trying commit 56795fb77a69bc61836537175564455067adada6 with merge 6faa3e6fd2949d5cb421d19522f597501fc751a7...
<!-- homu: {"type":"TryBuildStarted","head_sha":"56795fb77a69bc61836537175564455067adada6","merge_sha":"6faa3e6fd2949d5cb421d19522f597501fc751a7"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13102334597/job/36554651951)
Build commit: 6faa3e6fd2949d5cb421d19522f597501fc751a7 (`6faa3e6fd2949d5cb421d19522f597501fc751a7`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13102334597/job/36554651951"},"merge_sha":"6faa3e6fd2949d5cb421d19522f597501fc751a7"} -->

saethlin commented:
@bors r=workingjubilee

bors commented:
:pushpin: Commit 56795fb77a69bc61836537175564455067adada6 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 56795fb77a69bc61836537175564455067adada6 -->
<!-- homu: {"type":"Approved","sha":"56795fb77a69bc61836537175564455067adada6","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 56795fb77a69bc61836537175564455067adada6 with merge e28631908591cd8d8980357e575b2ce891500198...
<!-- homu: {"type":"BuildStarted","head_sha":"56795fb77a69bc61836537175564455067adada6","merge_sha":"e28631908591cd8d8980357e575b2ce891500198"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13104917584/job/36558122668) [(plain)](https://github.com/rust-lang-ci/rust/commit/e28631908591cd8d8980357e575b2ce891500198/checks/36558122668/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13104917584/job/36560918148)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13104917584/job/36560918148","builder_name":"checks-actions"} -->

tgross35 commented:
Loongarch failed, web logs just stopped and there are no text logs. I think this has to be a GH issue but might as well double check.

@bors try

bors commented:
:hourglass: Trying commit 56795fb77a69bc61836537175564455067adada6 with merge e50c9cb654efdc7ec36525adb0b29080e6cd6432...
<!-- homu: {"type":"TryBuildStarted","head_sha":"56795fb77a69bc61836537175564455067adada6","merge_sha":"e50c9cb654efdc7ec36525adb0b29080e6cd6432"} -->

tgross35 commented:
Crosslinking a zulip thread about this PR's CI misfortunes https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/Spooky.20loongarch64.20failures

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13106402415/job/36561892714) [(plain)](https://github.com/rust-lang-ci/rust/commit/e50c9cb654efdc7ec36525adb0b29080e6cd6432/checks/36561892714/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13106402415/job/36564879226)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13106402415/job/36564879226","builder_name":"checks-actions"} -->

tgross35 commented:
Same failure. Huh

tgross35 commented:
I don't think anybody has switched to the larger runner yet but it's been a few days, maybe we'll get lucky.

@bors try

bors commented:
:hourglass: Trying commit 56795fb77a69bc61836537175564455067adada6 with merge 99f1fc523e889feb39592a95b5baecaeb3af8e21...
<!-- homu: {"type":"TryBuildStarted","head_sha":"56795fb77a69bc61836537175564455067adada6","merge_sha":"99f1fc523e889feb39592a95b5baecaeb3af8e21"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13153882718/job/36706566255) [(plain)](https://github.com/rust-lang-ci/rust/commit/99f1fc523e889feb39592a95b5baecaeb3af8e21/checks/36706566255/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13153882718/job/36712705868)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13153882718/job/36712705868","builder_name":"checks-actions"} -->

Flakebi commented:
I opened a PR to switch the loongarch jobs to `largedisk` runners: #136679
When we try here again, we should include the two loongarch jobs and `dist-powerpc64-linux`, the powerpc one succeeded the last times, but also ran out of disk space once in one of the rollups.
I saw the PR description was edited for the try jobs, does it work to just add more jobs there? (I’m not sure on the syntax)

saethlin commented:
Yeah you just add more try-job: lines

Flakebi commented:
@bors try
Let’s see if it works with the larger runners

bors commented:
@Flakebi: :key: Insufficient privileges: not in try users

matthiaskrgr commented:
@bors try

bors commented:
:hourglass: Trying commit 56795fb77a69bc61836537175564455067adada6 with merge f78acd20fca5d7919d741bb3711814d7c332848e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"56795fb77a69bc61836537175564455067adada6","merge_sha":"f78acd20fca5d7919d741bb3711814d7c332848e"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13224478814/job/36913590254) [(plain)](https://github.com/rust-lang-ci/rust/commit/f78acd20fca5d7919d741bb3711814d7c332848e/checks/36913590254/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
dist-loongarch64-linux
dist-loongarch64-muls
dist-powerpc64-linux
##[endgroup]
INFO:root:Job type: TryRunType(custom_jobs=['dist-loongarch64-linux', 'dist-loongarch64-muls', 'dist-powerpc64-linux'])
  File "/home/runner/work/rust/rust/src/ci/github-actions/ci.py", line 314, in <module>
    calculate_job_matrix(data)
  File "/home/runner/work/rust/rust/src/ci/github-actions/ci.py", line 266, in calculate_job_matrix
  File "/home/runner/work/rust/rust/src/ci/github-actions/ci.py", line 266, in calculate_job_matrix
    jobs = calculate_jobs(run_type, job_data)
  File "/home/runner/work/rust/rust/src/ci/github-actions/ci.py", line 153, in calculate_jobs
    raise Exception(
    raise Exception(
Exception: Custom job(s) `['dist-loongarch64-muls']` not found in auto jobs
##[error]Process completed with exit code 1.
```

</details>

matthiaskrgr commented:
ah its called musl and not muls :) 
@bors try

bors commented:
:hourglass: Trying commit 56795fb77a69bc61836537175564455067adada6 with merge ad8d58687ff5e1b7935c4b25be4e251d15443948...
<!-- homu: {"type":"TryBuildStarted","head_sha":"56795fb77a69bc61836537175564455067adada6","merge_sha":"ad8d58687ff5e1b7935c4b25be4e251d15443948"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13224510086/job/36915365218)
Build commit: ad8d58687ff5e1b7935c4b25be4e251d15443948 (`ad8d58687ff5e1b7935c4b25be4e251d15443948`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13224510086/job/36915365218"},"merge_sha":"ad8d58687ff5e1b7935c4b25be4e251d15443948"} -->

saethlin commented:
@bors r=workingjubilee

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #133092

bors commented:
:pushpin: Commit 56795fb77a69bc61836537175564455067adada6 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 56795fb77a69bc61836537175564455067adada6 -->
<!-- homu: {"type":"Approved","sha":"56795fb77a69bc61836537175564455067adada6","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 56795fb77a69bc61836537175564455067adada6 with merge c03c38d5c2368cd2aa0e056dba060b94fc747f4e...
<!-- homu: {"type":"BuildStarted","head_sha":"56795fb77a69bc61836537175564455067adada6","merge_sha":"c03c38d5c2368cd2aa0e056dba060b94fc747f4e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13234087492/job/36941747238)
Approved by: workingjubilee
Pushing c03c38d5c2368cd2aa0e056dba060b94fc747f4e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"workingjubilee","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13234087492/job/36941747238"},"merge_sha":"c03c38d5c2368cd2aa0e056dba060b94fc747f4e"} -->

rust-timer commented:
Finished benchmarking commit (c03c38d5c2368cd2aa0e056dba060b94fc747f4e): [comparison URL](https://perf.rust-lang.org/compare.html?start=d9a4a47b8b3dc0bdff83360cea2013200d60d49c&end=c03c38d5c2368cd2aa0e056dba060b94fc747f4e&stat=instructions:u).

### Overall result: ❌ regressions - please read the text below


Our benchmarks found a performance regression caused by this PR.
This might be an actual regression, but it can also be just noise.

**Next Steps**:

- If the regression was expected or you think it can be justified,
please write a comment with sufficient written justification, and add
`@rustbot label: +perf-regression-triaged` to it, to mark the regression as triaged.
- If you think that you know of a way to resolve the regression, try to create
a new PR with a fix for the regression.
- If you do not understand the regression or you think that it is just noise,
you can ask the `@rust-lang/wg-compiler-performance` working group for help (members of this group
were already notified of this PR).

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d9a4a47b8b3dc0bdff83360cea2013200d60d49c&end=c03c38d5c2368cd2aa0e056dba060b94fc747f4e&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7% | [0.2%, 3.8%] | 55    |
| Regressions ❌ <br /> (secondary)  | 0.6% | [0.2%, 1.0%] | 47    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.7% | [0.2%, 3.8%] | 55    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d9a4a47b8b3dc0bdff83360cea2013200d60d49c&end=c03c38d5c2368cd2aa0e056dba060b94fc747f4e&stat=max-rss)
<details>
<summary>Results (primary 2.1%, secondary 3.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.2%  | [0.4%, 5.8%]   | 93    |
| Regressions ❌ <br /> (secondary)  | 3.4%  | [0.5%, 6.1%]   | 96    |
| Improvements ✅ <br /> (primary)   | -6.8% | [-6.8%, -6.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.5% | [-1.5%, -1.5%] | 1     |
| All ❌✅ (primary)                 | 2.1%  | [-6.8%, 5.8%]  | 94    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d9a4a47b8b3dc0bdff83360cea2013200d60d49c&end=c03c38d5c2368cd2aa0e056dba060b94fc747f4e&stat=cycles:u)
<details>
<summary>Results (primary 2.4%, secondary -2.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.4%  | [0.6%, 3.6%]   | 3     |
| Regressions ❌ <br /> (secondary)  | 2.4%  | [1.3%, 3.2%]   | 4     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -5.4% | [-9.7%, -4.0%] | 7     |
| All ❌✅ (primary)                 | 2.4%  | [0.6%, 3.6%]   | 3     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=d9a4a47b8b3dc0bdff83360cea2013200d60d49c&end=c03c38d5c2368cd2aa0e056dba060b94fc747f4e&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 777.876s -> 781.62s (0.48%)
**Artifact size**: 329.17 MiB -> 348.30 MiB (5.81%)

Kobzol commented:
The `syn` regression is noise, the rest is possibly caused by the fact that the shipped LLVM is now 19 MiB larger due to the inclusion of the AMDGPU target.

Kobzol commented:
I remember that sometimes binary/dynamic library size increases also increased icounts due to the dynamic linker doing more work. But based on CacheGrind, it looks like LLVM is actually doing more work, seems like it maybe iterates over more passes that were enabled by the amdgpu target?

lqd commented:
The max-rss increases also look unexpected, and numerous enough to not be measurement noise. (Could memory allocation be in these ??? cg reports, sometimes it does this for me rather than finding jemalloc/malloc)

nikic commented:
Adding the amdgpu target shouldn't make any additional passes run -- additional cost from registering additional passes etc is plausible though.

max-rss increasing with increasing code size is pretty common.
